Mathlib/Data/List/Count.lean

```diff
@@ -0,0 +1,392 @@
+  simp only [countp_eq_length_filter, filter_append, length_append]
+#align list.countp_append List.countp_append
+
+theorem countp_join : ∀ l : List (List α), countp p l.join = (l.map (countp p)).sum
+  | [] => rfl
+  | a :: l => by rw [join, countp_append, map_cons, sum_cons, countp_join l]
+#align list.countp_join List.countp_join
+
+theorem countp_pos : 0 < countp p l ↔ ∃ a ∈ l, p a := by
+  simp only [countp_eq_length_filter, length_pos_iff_exists_mem, mem_filter, exists_prop]
+#align list.countp_pos List.countp_pos
+
+@[simp]
+theorem countp_eq_zero : countp p l = 0 ↔ ∀ a ∈ l, ¬p a := by
+  rw [← not_iff_not, ← Ne.def, ← pos_iff_ne_zero, countp_pos]
+  simp
+#align list.countp_eq_zero List.countp_eq_zero
+
+@[simp]
+theorem countp_eq_length : countp p l = l.length ↔ ∀ a ∈ l, p a := by
+  rw [countp_eq_length_filter, filter_length_eq_length]
+#align list.countp_eq_length List.countp_eq_length
+
+theorem length_filter_lt_length_iff_exists (l) :
+    length (filter p l) < length l ↔ ∃ x ∈ l, ¬p x := by
+  simpa [length_eq_countp_add_countp p l, countp_eq_length_filter] using
+  countp_pos (fun x => ¬p x) (l := l)
+#align list.length_filter_lt_length_iff_exists List.length_filter_lt_length_iff_exists
+
+theorem Sublist.countp_le (s : l₁ <+ l₂) : countp p l₁ ≤ countp p l₂ := by
+  simpa only [countp_eq_length_filter] using length_le_of_sublist (s.filter p)
+#align list.sublist.countp_le List.Sublist.countp_le
+
+@[simp]
+theorem countp_filter (l : List α) : countp p (filter q l) = countp (fun a => p a ∧ q a) l := by
+  simp only [countp_eq_length_filter, filter_filter]
+#align list.countp_filter List.countp_filter
+
+@[simp]
+theorem countp_true : (l.countp fun _ => true) = l.length := by simp
+#align list.countp_true List.countp_true
+
+@[simp]
+theorem countp_false : (l.countp fun _ => false) = 0 := by simp
+#align list.countp_false List.countp_false
+
+@[simp]
+theorem countp_map (p : β → Bool) (f : α → β) :
+    ∀ l, countp p (map f l) = countp (p ∘ f) l
+  | [] => rfl
+  | a :: l => by rw [map_cons, countp_cons, countp_cons, countp_map p f l]; rfl
+#align list.countp_map List.countp_map
+
+variable {p q}
+
+theorem countp_mono_left (h : ∀ x ∈ l, p x → q x) : countp p l ≤ countp q l := by
+  induction' l with a l ihl
+  · rfl
+  rw [forall_mem_cons] at h
+  cases' h with ha hl
+  rw [countp_cons, countp_cons]
+  refine' _root_.add_le_add (ihl hl) _
+  split_ifs <;> simp only [le_rfl, _root_.zero_le]
+  exact absurd (ha ‹_›) ‹_›
+#align list.countp_mono_left List.countp_mono_left
+
+theorem countp_congr (h : ∀ x ∈ l, p x ↔ q x) : countp p l = countp q l :=
+  _root_.le_antisymm
+    (countp_mono_left fun x hx => (h x hx).1)
+    (countp_mono_left fun x hx => (h x hx).2)
+#align list.countp_congr List.countp_congr
+
+end Countp
+
+/-! ### count -/
+
+
+section Count
+
+variable [DecidableEq α]
+
+@[simp]
+theorem count_nil (a : α) : count a [] = 0 := rfl
+#align list.count_nil List.count_nil
```

Comment:
Linter claims that `simp` can prove this. But it's a good `dsimp`-lemma. Also, simp can only prove it using https://github.com/leanprover-community/mathlib4/pull/1410/files#diff-cdbce63b7556c5dc37babb0180e51282c7467545742bb9be6bc251f1db148b97R256 below.
Should we bump priority here?